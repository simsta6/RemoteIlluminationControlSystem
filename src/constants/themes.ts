import { Theme } from "@react-navigation/native";

export const KEY_FOR_STORAGE = "defaultScheme";

interface Colors {
    primary: string;
    background: string;
    card: string;
    text: string;
    disabledText: string;
    icon: string;
    border: string;
    notification: string;
    modal: string;
    button: string;
    disabledButton: string;
}

export const AppDarkColors: Colors = {
    primary: "rgba(133, 170, 164, 1)",
    background: "rgba(48, 48, 48, 1)",
    card: "rgba(33, 33, 33, 1)",
    text: "rgba(255, 228, 234, 1)", 
    disabledText: "rgba(186, 168, 172, 1)", 
    icon: "rgba(255, 228, 234, 1)", 
    border: "rgba(51, 70, 69, 1)",
    notification: "rgba(76, 96, 68, 1)",
    modal: "rgba(66, 66, 66, 1)",
    button: "rgba(33, 33, 33, 1)",
    disabledButton: "rgba(74, 74, 74, 1)",
};

export const AppLightColors: Colors = {
    primary: "rgba(122, 85, 91, 1)",
    background: "rgba(255, 235, 236, 1)",
    card: "rgba(229, 208, 237, 1)",
    text: "rgba(122, 85, 91, 1)", 
    disabledText: "rgba(87, 74, 76, 1)", 
    icon: "rgba(122, 85, 91, 1)", 
    border: "rgba(204, 185, 186, 1)",
    notification: "rgba(179, 159, 187, 1)",
    modal: "rgba(255, 247, 248, 1)",
    button: "rgba(229, 208, 237, 1)",
    disabledButton: "rgba(171, 163, 173, 1)",
};

export const AppDarkTheme: Theme = {
    dark: true,
    colors: AppDarkColors,
};

export const AppLightTheme: Theme = {
    dark: false,
    colors: AppLightColors
};