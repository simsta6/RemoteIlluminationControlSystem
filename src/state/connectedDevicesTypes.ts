import { Device, Subscription } from "react-native-ble-plx";

export interface DeviceState {
    device: Device,
    subscription?: Subscription,
    isDeviceConnected: boolean,
}

export type ConnectDeviceState = { 
    devices: DeviceState[]
}

export const ConnectedDevicesActionsTypes = {
    Add: "AddDevice",
    Remove: "RemoveDevice",
    Modify: "ModifyDevice",
} as const;

export type AddDeviceAction = {
    type: typeof ConnectedDevicesActionsTypes.Add;
    deviceState: DeviceState;
}

export type ModifyDeviceAction = {
    type: typeof ConnectedDevicesActionsTypes.Modify;
    index: number;
    deviceState: DeviceState;
}

export type RemoveDeviceAction = {
    type: typeof ConnectedDevicesActionsTypes.Remove;
    index: number;
}

export type ConnectedDevicesActions = AddDeviceAction | ModifyDeviceAction | RemoveDeviceAction;

