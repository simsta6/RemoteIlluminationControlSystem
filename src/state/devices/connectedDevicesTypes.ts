import { Subscription } from "react-native-ble-plx";

export interface DeviceState {
    device: Device,
    subscription?: Subscription,
    isDeviceConnected: boolean,
}

export interface Device {
    index: number;
    number: number;
    color: string;
    bulbType: "RGB" | "Non-RGB";
    temperature: string;
    power: string;
    voltage: string;
    current: string;
}

export type ConnectDeviceState = { 
    devices: Device[]
}

export const ConnectedDevicesActionsTypes = {
    Add: "AddDevice",
    Remove: "RemoveDevice",
    Modify: "ModifyDevice",
} as const;

export type AddDeviceAction = {
    type: typeof ConnectedDevicesActionsTypes.Add;
    deviceState: Device;
}

export type ModifyDeviceAction = {
    type: typeof ConnectedDevicesActionsTypes.Modify;
    index: number;
    deviceState: Device;
}

export type RemoveDeviceAction = {
    type: typeof ConnectedDevicesActionsTypes.Remove;
    index: number;
}

export type ConnectedDevicesActions = AddDeviceAction | ModifyDeviceAction | RemoveDeviceAction;

