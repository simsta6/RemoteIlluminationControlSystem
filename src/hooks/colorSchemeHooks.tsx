import { useTheme } from "@react-navigation/native";
import { AppDarkColors, AppLightColors } from "../constants/themes";

export const useAppColors = () => {
    const { dark } = useTheme();

    return {
        colors: dark ? AppDarkColors : AppLightColors
    };
};
