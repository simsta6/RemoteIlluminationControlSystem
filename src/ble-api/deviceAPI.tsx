import base64 from "base-64";
import { BleManager, Device as PlxDevice, Subscription } from "react-native-ble-plx";
import { BleDeviceActionTypes } from "../state/actions";
import { BleDevice } from "../state/ble-device/bleDeviceTypes";
import { Device } from "../state/devices/connectedDevicesTypes";
import { BLE_DEVICE_COMMANDS, generateMessage, ReadModuleParameter, TurnOnOrOffParameter } from "./messageGenerator";
import BluetoothStateManager from "react-native-bluetooth-state-manager";
import Toast from "react-native-toast-message";

type DevicesAdditionActionsType = {
    changeSvj: (newValue: number) => void;
    add: (device: Device) => void;
}

export class BleDeviceClient {
    private _bleManager: BleManager;
    private _bleDeviceId: string | undefined;
    private _actions: BleDeviceActionTypes;
    private _subscription: Subscription | undefined;
    private _serviceUUID: string | undefined;
    private _characteristicUUID: string | undefined;
    private _devicesAdditionActions: DevicesAdditionActionsType;
    private _tryingToConnect: boolean;

    public _didDeviceTriedToConnectOnStartup: boolean;

    constructor(bleManager: BleManager, bleDeviceId: string | undefined, serviceUUID: string, characteristicUUID: string, actions: BleDeviceActionTypes, devicesAdditionActions: DevicesAdditionActionsType) {
        this._bleManager = bleManager;
        this._bleDeviceId = bleDeviceId;
        this._actions = actions;
        this._devicesAdditionActions = devicesAdditionActions;
        this._subscription = undefined;
        this._serviceUUID = serviceUUID;
        this._characteristicUUID = characteristicUUID;

        this._tryingToConnect = false;
        this._didDeviceTriedToConnectOnStartup = false;
        bleDeviceId && this.connectToDevice(bleDeviceId).then(() => {
            this._didDeviceTriedToConnectOnStartup = true;
        });
    }

    public readResponse(command: string, bleDevice: BleDevice) {
        const response = bleDevice.messages.find(msg => msg.command.includes(command))?.response;
        this._actions.readResponse(command);
        return response;
    }

    public async connectToDevice(deviceId: string): Promise<boolean> {
        let connectedDevice: PlxDevice | undefined = undefined;
        const bleState = await BluetoothStateManager.getState();
        if (bleState === "PoweredOff") {
            await  BluetoothStateManager.requestToEnable();
        }

        if (!await this._bleManager.isDeviceConnected(deviceId) && !this._tryingToConnect) {
            this._tryingToConnect = true;
            connectedDevice = await this._bleManager
                .connectToDevice(deviceId)
                .then(cDevice => {
                    return cDevice.discoverAllServicesAndCharacteristics().then(async (deviceWithChar) => {
                        console.log("connected");
                        this._bleDeviceId = deviceWithChar.id;
                        await this.setPrimaryCharacteristic(this._bleManager, deviceWithChar.id);
                        await this.listenToMessages(deviceWithChar.id);
                        this._tryingToConnect = false;
                        return deviceWithChar;
                    });
                })
                .catch(err => {
                    this.showError(err);
                    return undefined;
                });
        }

        connectedDevice && this._actions.modify({
            deviceId: connectedDevice.id,
            isDeviceConnected: !!(connectedDevice.id && this._subscription),
            messages: [],
            serviceUUID: this._serviceUUID,
            uuid: this._characteristicUUID,
        });

        return connectedDevice ? true : false;
    }

    public async sendMessage(message: string, needResponse = true, timesToTry = 2): Promise<boolean> {
        const characteristicsUUID = this._characteristicUUID;
        const serviceUUID = this._serviceUUID;
        const bleDeviceId = this._bleDeviceId;
        if (!characteristicsUUID || !serviceUUID || !bleDeviceId) return false;

        if (await this._bleManager.isDeviceConnected(bleDeviceId) && this._subscription) {
            await this._bleManager.writeCharacteristicWithResponseForDevice(bleDeviceId, serviceUUID, characteristicsUUID, base64.encode(message + "\r\n"))
                .then((char) => {
                    if (char.value) {
                        const command = base64.decode(char.value);
                        needResponse && this._actions.sendRequest(command);
                    }
                })
                .catch(err => {
                    this.showError(err);
                    return false;
                });
            return true; // to prevent sending message if this client did not tried to connect to device 
        } else {
            if (timesToTry < 1) return false;
            await this.connectToDevice(bleDeviceId);
            const res = await this.sendMessage(message, needResponse, timesToTry - 1);
            return res;
        }
    }

    public get bleManager() {
        return this._bleManager;
    }

    public get didDeviceTriedToConnectOnStartup() {
        return this._didDeviceTriedToConnectOnStartup;
    }

    public get bleDeviceId() {
        return this._bleDeviceId;
    }

    public async requestStats(): Promise<boolean> {
        const message = generateMessage(BLE_DEVICE_COMMANDS.GetStats);
        return await this.sendMessage(message, false);
    }

    public async sendMessageToGetModuleColor(moduleId: string): Promise<boolean> {
        const message = generateMessage(BLE_DEVICE_COMMANDS.ReadModule, moduleId, ReadModuleParameter.GetRgbColor);
        const isSuccessful = await this.sendMessage(message);
        return isSuccessful;
    }

    public async changeDeviceColorOrBrightness(moduleIdOrIds: string[], colors: { [x: string]: string; }): Promise<boolean> {
        const result = await Promise.all(moduleIdOrIds.map(async id => {
            const color = colors[id];
            const message = generateMessage(BLE_DEVICE_COMMANDS.ChangeColorOrBrightness, id, color.substring(1, 7));
            return await this.sendMessage(message, false);
        }));
        return !result.some(res => !res);
    }

    public async turnOnSpecificDevice(moduleIdOrIds: string[]): Promise<boolean> {
        const result = await Promise.all(moduleIdOrIds.map(async id => {
            const message = generateMessage(BLE_DEVICE_COMMANDS.TurnOnOrOffModule, id, TurnOnOrOffParameter.ON);
            return await this.sendMessage(message, false);
        }));
        return !result.some(res => !res);
    }

    public async turnOffSpecificDevice(moduleIdOrIds: string[]): Promise<boolean> {
        const result = await Promise.all(moduleIdOrIds.map(async id => {
            const message = generateMessage(BLE_DEVICE_COMMANDS.TurnOnOrOffModule, id, TurnOnOrOffParameter.OFF);
            return await this.sendMessage(message, false);
        }));
        return !result.some(res => !res);
    }

    public async turnOnAllDevices(): Promise<boolean> {
        const message = generateMessage(BLE_DEVICE_COMMANDS.TurnOnAllModules);
        return await this.sendMessage(message, false);
    }

    public async turnOffAllDevices(): Promise<boolean> {
        const message = generateMessage(BLE_DEVICE_COMMANDS.TurnOffAllModules);
        return await this.sendMessage(message, false);
    }

    public async testDeviceByBlinking(moduleId: string, initColor: string): Promise<boolean> {
        await this.sendMessage(generateMessage(BLE_DEVICE_COMMANDS.ChangeColorOrBrightness, moduleId, "050505"), false);
        await new Promise(r => setTimeout(r, 500));
        const result =  await this.sendMessage(generateMessage(BLE_DEVICE_COMMANDS.ChangeColorOrBrightness, moduleId, initColor.substring(1, 7)), false);
        return result;
    }

    public async changeLightSensorValue(newValue: number): Promise<boolean> {
        const message = generateMessage(BLE_DEVICE_COMMANDS.ChangeLightSensorValue, undefined, Math.round(newValue).toString(16));
        return await this.sendMessage(message, false);
    }

    private async listenToMessages(deviceId: string) {
        this._subscription?.remove();
        const characteristicsUUID = this._characteristicUUID;
        const serviceUUID = this._serviceUUID;

        if (characteristicsUUID && serviceUUID) {
            this._subscription = await this._bleManager.isDeviceConnected(deviceId)
                .then(isConnected => {
                    if (isConnected) {
                        return this._bleManager.monitorCharacteristicForDevice(deviceId, serviceUUID, characteristicsUUID, (error, char) => {
                            if (error) {
                                if (`${error.name}: ${error.message}`.includes("BleError: Device") && error.message.includes("was disconnected")
                                    || `${error.name}: ${error.message}`.includes("BleError: Characteristic") && error.message.includes("notify change failed for device")) {
                                    return;
                                }

                                this.showError(error.message);
                                return;
                            }
                            if (char?.value) {
                                const response = base64.decode(char.value).replace(/(\r\n|\n|\r)/gm,"");
                                const responseArray = response.split(":");
                                if (responseArray.length > 1) {
                                    const command = responseArray[0];
                                    const response = responseArray.slice(1);
                                    if (command.includes(BLE_DEVICE_COMMANDS.GetStats)) {
                                        const values = response.join(",").split(",");
                                        if (values[0].includes("SVJ")) {
                                            this._devicesAdditionActions.changeSvj(parseInt(values[2], 16));
                                        } else {
                                            const index = values[0].slice(2);
                                            const color = "#" + values[1].toUpperCase();
                                            const current = values[5] === "0" ? 0.045 : 0.0045;
                                            const voltage = parseInt(values[3], 16) / 1000;
                                            const power = current * voltage;
                                            const temperature = parseInt(values[4], 16).toString();
                                            const bulbType = values[5] === "0" ? "Non-RGB" : "RGB";
                                            
                                            this._devicesAdditionActions.add({
                                                index,
                                                color,
                                                bulbType,
                                                temperature,
                                                power: power.toString(),
                                                voltage: voltage.toString(),
                                                current: current.toString(),
                                            });
                                        }
                                    } else {
                                        this._actions.updateResponse(command, response);
                                    }
                                }

                                console.log(response);

                            }
                        });
                    }
                })
                .catch(err => {
                    this.showError(err);
                    return undefined;
                });
        }
    }

    private async setPrimaryCharacteristic(bleManager: BleManager, deviceId: string): Promise<boolean> {
        const services = await bleManager.servicesForDevice(deviceId);

        for (const service of services) {
            const characteristics = await bleManager.characteristicsForDevice(deviceId, service.uuid);

            for (const characteristic of characteristics) {
                if (characteristic.isWritableWithResponse) {
                    this._serviceUUID = characteristic.serviceUUID;
                    this._characteristicUUID = characteristic.uuid;
                    return true;
                }
            }
        }
        return false;
    }

    public async sendMessageWithValidation(message: string): Promise<boolean> {
        if (this.validateCommand(message)) {
            return this.sendMessage(message, false);
        }
        return Promise.resolve(false);
    }

    private validateCommand(command: string): boolean {
        const regexExpressions = [
            "LIST",
            "STATS",
            "ON",
            "OFF",
            "OK",
            "ID[0-9A-F]{1,2}CLR[0-9A-F]{6}",
            "ID[0-9A-F]{1,2}ST[1-5]",
            "ID[0-9A-F]{1,2}ONF[0-1]",
            "LX[0-9A-F]{1,3}"
        ] as const;

        for(let i = 0; i < regexExpressions.length; i++) {
            const re = new RegExp(regexExpressions[i]);
            if (re.exec(command)) {
                return true;
            }
        }

        return false;
    }

    private showError(err: string) {
        Toast.show({
            type: "error",
            text1: "Error",
            text2: err
        });
    }
}