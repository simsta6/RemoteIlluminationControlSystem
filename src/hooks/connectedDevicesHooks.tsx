import { useDispatch, useSelector } from "react-redux";
import { ConnectDevicesActions } from "../state/actions";
import { Device } from "../state/devices/connectedDevicesTypes";
import { ActionsUnion, IRootState } from "../state/types";

export const useConnectedDevices = () => {
    const { devices } = useSelector((state: IRootState) => state.devicesReducer);
    const dispatch: React.Dispatch<ActionsUnion<typeof ConnectDevicesActions>> = useDispatch();

    return [
        devices,
        {
            add: (device: Device) => 
                dispatch(ConnectDevicesActions.AddDevice(device)),
            remove: (index: number) => 
                dispatch(ConnectDevicesActions.RemoveDevice(index)),
            modify: (device: Device, index: number) => 
                dispatch(ConnectDevicesActions.ModifyDevice(device, index))
        }
    ] as const;
};