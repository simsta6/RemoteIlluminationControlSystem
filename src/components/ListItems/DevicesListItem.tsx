import React from "react";
import { StyleSheet, Text, View } from "react-native";
import ThreeVerticalDotsIcon from "../../assets/icons/ThreeVerticalDotsIcon";
import { useAppColors } from "../../hooks/colorSchemeHooks";
import { Device } from "../../state/devices/connectedDevicesTypes";
import { IconButton } from "../Buttons/IconButton";

interface Props {
    device: Device;
    Icon?: JSX.Element;
    isLast: boolean;
    setIsModalVisible: React.Dispatch<React.SetStateAction<boolean>>;
}

export const DevicesListItem = (props: Props) => {
    const { colors } = useAppColors();
    const { device, Icon, isLast, setIsModalVisible } = props;

    return (
        <>
            <View style={styles.row}>
                <View style={styles.closeItems}>
                    {Icon ? Icon : <></>}
                    <Text style={{ ...styles.deviceName, color: colors.text }}>{device.name}</Text>
                </View>
                <View style={{...styles.closeItems, ...styles.threeDotsContainer}}>
                    <IconButton 
                        Icon={() => <ThreeVerticalDotsIcon color={colors.text} height={25} width={25} />} 
                        onPress={() => setIsModalVisible(true)}
                    />
                </View>
            </View>
            {!isLast && <View style={{...styles.horizontalSeparator, backgroundColor: colors.text }}/>}
        </>
    );
};

const styles = StyleSheet.create({
    row: {
        flexDirection: "row",
        justifyContent: "space-between",
        alignItems: "center",
        height: 50,
    },
    closeItems: {
        flexDirection: "row",
        height: "100%",
        justifyContent: "flex-start",
        alignItems: "center",
    },
    horizontalSeparator: {
        width: "100%",
        height: 1,
        opacity: 0.3,
        marginVertical: 5,
    },
    deviceName: {
        fontWeight: "500",
        fontSize: 16,
        marginLeft: 16,
    },
    threeDotsContainer: {
        marginRight: 10,
    }
});
