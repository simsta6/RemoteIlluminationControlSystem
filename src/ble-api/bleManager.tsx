import React from "react";
import { BleManager, Device } from "react-native-ble-plx";
import BluetoothStateManager from "react-native-bluetooth-state-manager";
import { requestPermissions } from "../helpers/permissions";
import { useBleDevice } from "../hooks/bleDeviceHook";
import { useConnectedDevices } from "../hooks/connectedDevicesHooks";
import { BleDeviceClient } from "./deviceAPI";

export const useBleDeviceClient = () => {
    const [bleDevice, actions] = useBleDevice();
    const [, devicesActions] = useConnectedDevices();
    const bleDeviceClient = React.useMemo(() => 
        new BleDeviceClient(new BleManager(), bleDevice.deviceId, bleDevice.serviceUUID ?? "", bleDevice.uuid ?? "", actions, devicesActions.add), 
    []);
    return bleDeviceClient;
};

const startBleScan = (
    bleManager: BleManager,
    startScan: boolean,
    setIsScanStarted: React.Dispatch<React.SetStateAction<boolean>>,
    setAvailableBleDevices: React.Dispatch<React.SetStateAction<Device[]>>,
    knownDeviceId: string | undefined,
    setIsModalVisible: React.Dispatch<React.SetStateAction<boolean>>
) => {
    if (!startScan) {
        bleManager.stopDeviceScan();
        setIsScanStarted(false);
        console.log("stopping scan");
        return;
    }

    bleManager.startDeviceScan(null, null, async (err, device) => {            
        if (err) { 
            if (err.message.includes("BluetoothLE is powered off")) {
                return BluetoothStateManager.requestToEnable().then(() => startBleScan(bleManager, startScan, setIsScanStarted, setAvailableBleDevices, knownDeviceId, setIsModalVisible));
            } 
            console.log(err);
            return;
        }

        if (knownDeviceId) {
            const isDeviceConnected = await bleManager.isDeviceConnected(knownDeviceId);
            if (isDeviceConnected) {
                setIsScanStarted(false);
                setIsModalVisible(false);
                return;
            }
        }

        setIsScanStarted(true);

        if (device && device.name === "TTBG") {
            setAvailableBleDevices(devices => devices.some(dev => dev.id === device.id) ? devices : [...devices, device]);
        }
    });
};

export const useScannedDevices = (
    bleManager: BleManager,
    setAvailableBleDevices: React.Dispatch<React.SetStateAction<Device[]>>, 
    startScan: boolean,
    knownDeviceId: string | undefined,
    setIsModalVisible: React.Dispatch<React.SetStateAction<boolean>>
) => {
    const [isScanStarted, setIsScanStarted] = React.useState(false);
    const [isPermissionsGranted, setIsPermissionsGranted] = React.useState(false);

    React.useEffect(() => {
        let setHook = true;

        requestPermissions().then(res => setHook && setIsPermissionsGranted(res));

        return () => {
            setHook = false;
        };
    }, []);

    React.useEffect(() => {
        if (isPermissionsGranted) 
            startBleScan(bleManager, startScan, setIsScanStarted, setAvailableBleDevices, knownDeviceId, setIsModalVisible);
    }, [bleManager, startScan, isPermissionsGranted]);

    return { isScanStarted };
};
