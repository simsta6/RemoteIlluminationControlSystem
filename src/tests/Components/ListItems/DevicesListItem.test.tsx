import "@testing-library/jest-dom";
import { fireEvent, render, waitFor } from "@testing-library/react-native";
import React from "react";
import { Text } from "react-native";
import { BleDeviceClient } from "../../../ble-api/deviceAPI";
import { DevicesListItem } from "../../../components/ListItems/DevicesListItem";
import { Colors } from "../../../constants/themes";
import * as hooks from "../../../hooks/colorSchemeHooks";
import * as connectedDevicesHooks from "../../../hooks/connectedDevicesHooks";
import { Device } from "../../../state/devices/connectedDevicesTypes";

describe("<DevicesListItem />", () => {
    const colors = {
        text: "purple",
    };

    const Icon = <Text>{"MockedIcon"}</Text>;

    let mockTestDeviceByBlinking: jest.SpyInstance<Promise<boolean>, [moduleId: string, initColor: string]>;

    const mockedBleDeviceClient = BleDeviceClient.prototype;

    beforeEach(() => {
        mockTestDeviceByBlinking = jest
            .spyOn(BleDeviceClient.prototype, "testDeviceByBlinking")
            .mockImplementation(() => Promise.resolve(true));

        jest.spyOn(BleDeviceClient.prototype, "didDeviceTriedToConnectOnStartup", "get").mockImplementation( () => true);

        spyOn(console, "error");
        
        jest.spyOn(hooks, "useAppColors").mockImplementation(() => ({ colors: colors as Colors }));

        jest.spyOn(connectedDevicesHooks, "useConnectedDevices").mockImplementation(() => [[{name: ""} as Device], {
            add: () => ({}),
            remove: () => ({}),
            removeAll: () => ({}),
            changeColor: () => ({}),
            modify: () => ({}),
        }]);
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    test("Default ConnectToDeviceListItem should match snapshot", () => {
        const props = {
            device: {
                index: "1",
                color: "#ff0000",
                name: "Test Device"
            } as Device,
            deviceIndexInArray: 0,
            Icon,
            isLast: false,
            bleDeviceClient: mockedBleDeviceClient,
        };

        const { toJSON } = render(
            <DevicesListItem {...props} />
        );

        expect(toJSON()).toMatchSnapshot();
    });

    test("should generate horizontal separator if isLast = false", async () => {
        const props = {
            device: {
                index: "1",
                color: "#ff0000",
                name: "Test Device"
            } as Device,
            deviceIndexInArray: 0,
            Icon,
            isLast: false,
            bleDeviceClient: mockedBleDeviceClient,
        };

        const { queryAllByTestId } = render(
            <DevicesListItem {...props} />
        );

        await waitFor(() => {
            expect(queryAllByTestId("horizontalSeparator")).toHaveLength(1);
        });
    });

    test("should not generate horizontal separator if isLast = true", async () => {
        const props = {
            device: {
                index: "1",
                color: "#ff0000",
                name: "Test Device"
            } as Device,
            deviceIndexInArray: 0,
            Icon,
            isLast: true,
            bleDeviceClient: mockedBleDeviceClient,
        };

        const { queryAllByTestId } = render(
            <DevicesListItem {...props} />
        );


        await waitFor(() => {
            expect(queryAllByTestId("horizontalSeparator")).toHaveLength(0);
        });
    });

    test("should call testDeviceByBlinking when custom Icon is pressed", async () => {
        const props = {
            device: {
                index: "1",
                color: "#ff0000",
                name: "Test Device"
            } as Device,
            deviceIndexInArray: 0,
            Icon,
            isLast: true,
            bleDeviceClient: mockedBleDeviceClient,
        };

        const { getByText } = render(
            <DevicesListItem {...props} />
        );


        fireEvent.press(getByText("MockedIcon"));
        await waitFor(() => expect(mockTestDeviceByBlinking).toHaveBeenCalledTimes(1));
    });

    test("should render Modal when ThreeVerticalDots Icon is pressed", async () => {
        const props = {
            device: {
                index: "1",
                color: "#ff0000",
                name: "Test Device"
            } as Device,
            deviceIndexInArray: 0,
            Icon,
            isLast: true,
            bleDeviceClient: mockedBleDeviceClient,
        };

        const { getByTestId, queryAllByTestId } = render(
            <DevicesListItem {...props} />
        );


        fireEvent.press(getByTestId("ThreeVerticalDotsIcon"));
        await waitFor(() => expect(queryAllByTestId("ModalCenteredView")).toHaveLength(1));
    });
});