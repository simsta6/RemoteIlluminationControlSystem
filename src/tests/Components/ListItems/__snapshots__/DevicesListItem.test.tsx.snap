// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DevicesListItem /> Default ConnectToDeviceListItem should match snapshot 1`] = `
Array [
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "height": 50,
        "justifyContent": "space-between",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "height": "100%",
          "justifyContent": "flex-start",
        }
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="TouchableOpacityIconButton"
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
            }
          }
        >
          <Text>
            MockedIcon
          </Text>
        </View>
      </View>
      <Text
        style={
          Object {
            "color": "purple",
            "fontSize": 16,
            "fontWeight": "500",
            "marginLeft": 16,
          }
        }
      >
        Test Device
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "height": "100%",
          "justifyContent": "flex-start",
          "marginRight": 10,
        }
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="TouchableOpacityIconButton"
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
            }
          }
        >
          <RNSVGSvgView
            align="xMidYMid"
            bbHeight={25}
            bbWidth={25}
            focusable={false}
            height={25}
            meetOrSlice={0}
            minX={0}
            minY={0}
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "borderWidth": 0,
                },
                Object {
                  "flex": 0,
                  "height": 25,
                  "width": 25,
                },
              ]
            }
            testID="ThreeVerticalDotsIcon"
            vbHeight={16}
            vbWidth={16}
            width={25}
          >
            <RNSVGGroup>
              <RNSVGPath
                d="M9.5 13a1.5 1.5 0 11-3 0 1.5 1.5 0 013 0zm0-5a1.5 1.5 0 11-3 0 1.5 1.5 0 013 0zm0-5a1.5 1.5 0 11-3 0 1.5 1.5 0 013 0z"
                fill={4286578816}
                propList={
                  Array [
                    "fill",
                  ]
                }
              />
            </RNSVGGroup>
          </RNSVGSvgView>
        </View>
      </View>
    </View>
  </View>,
  <View
    style={
      Object {
        "backgroundColor": "purple",
        "height": 1,
        "marginVertical": 5,
        "opacity": 0.3,
        "width": "100%",
      }
    }
    testID="horizontalSeparator"
  />,
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    transparent={true}
    visible={false}
  />,
]
`;
